### DISCUSSION TITLE
What's preventing the use of GPU in docker containers?

### DISCUSSION THREAD

**Discussion by AnkushMalaker** [khoj-ai/khoj#854] [Q&A]
I was able to give access to my GPU by modifying the docker-compose file and using `nvidia/cuda:12.1.1-runtime-ubuntu22.04` as the base image. 
torch.cuda.is_available() is also True.
What's preventing the LLM from utilizing the GPU that'd work otherwise with local install, according to the [offline + gpu instructions](https://docs.khoj.dev/get-started/setup#setup)?

**Comment by debanjum** [comment]
That's awesome to hear! What OS are you running on? And do you see GPU utilization spike when offline chat model responds?

Few things that needed to be done were:
1. Test offline chat models utilize GPU on Ubuntu, Windows, Mac when using Khoj docker image built on Nvidia base
2. Add/update Dockerfile, Github workflow to publish (separate?) Nvidia based Khoj docker images, once step 1 works

**Reply by debanjum** [reply]
The main thing to verify was that the underlying library we use, llama.cpp, is able to see+use the system GPU on the major operating systems when used via Docker

**Reply by AnkushMalaker** [reply]
I'm running windows with docker on wsl backend with RTX 4090. 
No, no spike. All processors were 100% indicating full CPU use.

**Reply by AnkushMalaker** [reply]
I'm pretty sure llama.cpp can see and use the system GPU. 
And I think I can give it a shot to try to run it on Docker on windows which is major enough? 
I also have an Ubuntu system so I can see about it running on both which is also super important to me lol

**Reply by AnkushMalaker** [reply]
Point me to where int the code we make decisions about where to run the LLM? (ie CPU vs GPU)

**Reply by rehandaphedar** [reply]
It is https://github.com/khoj-ai/khoj/blob/892040972fed8b6236c4ad4180f992707a024508/src/khoj/utils/helpers.py#L297

**Comment by rehandaphedar** [comment]
Something information that might be useful - I used another program [fish](https://speech.fish.audio/#docker-setup) which supports using GPU for training+inference in Docker.

It seems to come down to:
- [Installing NVidia Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html)
- Passing `--gpus all` when running Docker

I tried doing the same thing. However, `torch.cuda.is_available()` returned False.
